outrank/algorithms/synthetic_data_generators/cc_generator.py

```diff
@@ -0,0 +1,849 @@
+
+                    for i in range(1, len(percentiles) - 1):
+                        percentiles[i] += percentiles[i - 1]
+
+                    percentiles.insert(0, 0)
+                    percentiles.pop()
+                    print(percentiles)
+
+                    p_points = np.percentile(decision_boundary, percentiles)
+                    print(p_points)
+
+                    y = np.zeros_like(decision_boundary, dtype=int)
+                    for i in range(1, n):
+                        p_point = p_points[i]
+                        for j in range(len(decision_boundary)):
+                            if decision_boundary[j] > p_point:
+                                y[j] += 1
+            else:
+                decision_boundary = decision_function(X)
+                p_point = np.percentile(decision_boundary, p * 100)
+                y = np.where(decision_boundary > p_point, 1, 0)
+        else:
+            if p == 0.5:
+                p = 1.0
+            else:
+                p = [p, 1 - p]
+            y = self._cluster_data(X, n, p=p, balance=balance)
+
+        self.dataset_info.update({
+            'labels': {
+                'class_relation': class_relation,
+                'n_class': n,
+            },
+        })
+
+        return y
+
+    def _cluster_data(
+        self,
+        X: ArrayLike,
+        n: int,
+        p: float | list[float] | ArrayLike | None = 1.0,
+        balance: bool = False,
+    ) -> np.ndarray:
+        """
+        Cluster data using kmeans
+        :param X: dataset
+        :param n: number of clusters
+        :param p: class distribution
+        :param balance: balance the clusters according to p
+        :return: array of labels, corresponding to dataset X
+        """
+
+        kmeans = KMeans(n_clusters=n)
+
+        kmeans.fit(X)
+
+        cluster_labels = kmeans.labels_
+
+        if not isinstance(p, (list, np.ndarray)):  # Fully balanced clusters
+            samples_per_cluster = [len(X) // n] * n
+        else:
+            samples = len(X)
+            samples_per_cluster = []
+            if not isinstance(p, (list, np.ndarray)):
+                samples_per_cluster.append(int(samples * p) // n)
+                samples_per_cluster.append(int(samples * (1 - p)) // n)
+            else:
+                if len(p) == n:
+                    for val in p:
+                        samples_per_cluster.append(int(samples * val))
+                else:
+                    raise Exception('Length of balance parameter must equal number of clusters.')
+
+        # Adjust cluster sizes
+        if balance:
+            adjustments = []
+            overflow_samples = []
+            overflow_indices = []
+            for i in range(n):
+                cluster_size = np.sum(cluster_labels == i)
+
+                adjustment = samples_per_cluster[i] - cluster_size
+                adjustments.append(adjustment)
+
+                if adjustment < 0:  # Cluter is too large
```

Comment:
Comments always above lines, not after pls

Reply:
Moved all comments above code block they're referencing.